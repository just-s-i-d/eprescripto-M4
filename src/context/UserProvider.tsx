import {
  PropsWithChildren,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from "react";

import { getUserData } from "@utils/Doctor";
import { showToast } from "@utils/common";

type UserContextType = {
  isLoggedIn?: boolean;
  setIsLoggedIn?: React.Dispatch<SetStateAction<boolean>>;
  isDarkTheme?: boolean;
  setIsDarkTheme?: React.Dispatch<SetStateAction<boolean>>;
  role?: string;
  setRole?: React.Dispatch<SetStateAction<string | undefined>>;
  refresh?: boolean;
  setRefresh?: React.Dispatch<SetStateAction<boolean | undefined>>;
  getRole?: () => void;
};

export const UserContext = createContext<UserContextType>({});
export const UserProvider = ({ children }: PropsWithChildren) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [isDarkTheme, setIsDarkTheme] = useState<boolean>(false);
  const [role, setRole] = useState<string>();

  const getRole = () => {
    getUserData()
      .then((res) => {
        setRole(res.role?.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      });
  };
  useEffect(() => {
    const token = localStorage.getItem("token");
    if (token) {
      getRole();
    }
    setIsLoggedIn(!!token);
  }, []);
  const value = {
    isLoggedIn,
    setIsLoggedIn,
    role,
    setRole,
    isDarkTheme,
    setIsDarkTheme,
    getRole,
  };
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};
