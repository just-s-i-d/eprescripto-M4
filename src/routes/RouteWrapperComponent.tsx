import { ReactElement, useContext } from "react";

import Loading from "@components/ui/Loading";
import { UserContext } from "@context/UserProvider";

type RouteWrapperComponentProps = {
  components: { [key: string]: ReactElement };
};
const RouteWrapperComponent = ({ components }: RouteWrapperComponentProps) => {
  const userContext = useContext(UserContext);
  const role = userContext?.role;
  return role ? components?.[role] : <Loading />;
};
export default RouteWrapperComponent;
