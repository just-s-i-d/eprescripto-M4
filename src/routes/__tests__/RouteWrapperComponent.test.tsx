import { UserContext } from "@context/UserProvider";
import RouteWrapperComponent from "@routes/RouteWrapperComponent";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
type WrapperComponentProps = {
  role?: string;
};
const mockedComponents = {
  doctor: <h1>Doctor</h1>,
};
const WrapperComponent = ({ role }: WrapperComponentProps) => {
  return (
    <UserContext.Provider value={{ role: role }}>
      <RouteWrapperComponent components={mockedComponents} />
    </UserContext.Provider>
  );
};
describe("Route Wrapper Component tests", () => {
  it("Snapshot test", () => {
    const { asFragment } = render(<WrapperComponent role="doctor" />);
    expect(screen.getByText("Doctor")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Loading test", () => {
    render(<WrapperComponent />);
    expect(screen.getByAltText("loader")).toBeInTheDocument();
  });
});
