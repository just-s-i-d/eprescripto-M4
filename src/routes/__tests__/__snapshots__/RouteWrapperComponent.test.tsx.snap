// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route Wrapper Component tests Snapshot test 1`] = `
<DocumentFragment>
  <h1>
    Doctor
  </h1>
</DocumentFragment>
`;
